### PR TITLE
SIMSBIOHUB-106: Saving regions on project/ survey create/ update

### PR DESCRIPTION
**Pull request by al-rosenthal**
# Overview

## Links to Jira tickets

[SIMSBIOHUB-106](https://apps.nrs.gov.bc.ca/int/jira/secure/RapidBoard.jspa?rapidView=8685&projectKey=SIMSBIOHUB&view=detail&selectedIssue=SIMSBIOHUB-106)

## Description of relevant changes

- Added new service and repo to facilitate region saving
- Modified project create/ edit functions to save regions
- Modified survey create/ edit functions to save regions
- Updated Project list to fill region column

## PR Testing
Regions on a project
1. Create a new project
2. Navigate to project list page 
3. See region column filled for your new project
4. Edit newly created project
5. Update region to a new region 
6. Navigate back to project list page
7. See updated region for your new project

Regions on a survey
1. Create a new survey 
2. Add a region while creating survey
3. Open DBeaver and check the `survey_region` table for new entries related to the newly created survey
4. Edit the survey and update to a new region
5. Open DBeaver and check the `survey_region` table for new entries related to the newly created survey

## PR Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

### CODE REVIEW

File: api/src/services/region-service.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+  /**
+   * Gets a unique list of region details for a given list of features
+   *
+   * @param {Feature[]} features
+   * @returns {*} {Promise<RegionDetails[]>}
+   */
+  async getUniqueRegionsForFeatures(features: Feature[]): Promise<RegionDetails[]> {
+    const bcgwService = new BcgwLayerService();
+    return await bcgwService.getUniqueRegionsForFeatures(features, this.connection);
```
Comment (curtisupshall) [inline api/src/services/region-service.ts]: Nit: change `return await` instances to just `return`

File: api/src/services/region-service.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+   *
+   * @param {number} projectId
+   * @param {IRegion[]} regions
+   */
+  async addRegionsToProject(projectId: number, regions: IRegion[]): Promise<void> {
+    // remove existing regions from a project
+    this.regionRepository.deleteRegionsFromAProject(projectId);
+
+    const regionIds = regions.map((item) => item.region_id);
+    await this.regionRepository.addRegionsToAProject(projectId, regionIds);
```
Comment (curtisupshall) [inline api/src/services/region-service.ts]: Nit: rename to `addRegionsToProject`

File: app/src/features/projects/list/ProjectsListPage.tsx [hunk 3]
```diff
@@ -134,7 +134,7 @@ const ProjectsListPage: React.FC = () => {
           <Paper elevation={0}>
             <Toolbar style={{ display: 'flex', justifyContent: 'space-between' }}>
               <Typography variant="h4" component="h2">
-                Projects found{' '}
```
Comment (curtisupshall) [inline app/src/features/projects/list/ProjectsListPage.tsx:137]: Can replace `{' '}` with a space, followed by `&zwnj` (zero-width non-joiner):

```tsx
<Typography variant="h4" component="h2">
  Projects found &zwnj
  <Typography className={classes.toolbarCount} component="span" variant="inherit" color="textSecondary">
    ({projectsDataLoader.data?.length || 0})
  </Typography>
</Typography>
````

File: api/src/services/region-service.ts (hunk 2 above)
Comment (al-rosenthal) [reply api/src/services/region-service.ts]: done

File: api/src/services/region-service.ts (hunk 1 above)
Comment (al-rosenthal) [reply api/src/services/region-service.ts]: done

File: app/src/features/projects/list/ProjectsListPage.tsx (hunk 3 above)
Comment (al-rosenthal) [reply app/src/features/projects/list/ProjectsListPage.tsx:137]: replaced

**Review by curtisupshall — DISMISSED** [review]
Code looks good!

File: api/src/paths/project/{projectId}/update.ts [hunk 4]
```diff
@@ -432,7 +433,7 @@ export interface IUpdateProject {
   coordinator: object | null;
   project: object | null;
   objectives: object | null;
-  location: object | null;
+  location: { geometry: Feature[] } | null;
```
Comment (NickPhura) [inline api/src/paths/project/{projectId}/update.ts]: Doesn't this also include the location description field? Unless that is somewhere else (under project maybe?)

File: api/src/repositories/region-repository.ts [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      }
+    });
+
+    sql.append(';');
+
+    try {
+      await this.connection.sql(sql);
+    } catch (error) {
+      throw new ApiExecuteSQLError('Failed to execute insert SQL for project_region', [
+        'RegionRepository->addRegionsToAProject'
```
Comment (NickPhura) [inline api/src/repositories/region-repository.ts]: nit: old function name? Same with a few other errors in this file. Looks like its mostly just the 'A'

Comment (al-rosenthal) [reply api/src/repositories/region-repository.ts]: updated

File: api/src/paths/project/{projectId}/update.ts (hunk 4 above)
Comment (al-rosenthal) [reply api/src/paths/project/{projectId}/update.ts]: yes it does, updated

**Review by NickPhura — APPROVED** [review]

**Review by curtisupshall — APPROVED** [review]

**Pull request merged**